src/GraphQL/Server.hs

```diff
@@ -248,24 +258,52 @@ instance forall f fs m.
 
   runFields _ f = queryError ("Unexpected Selection value. Is the query normalized?: " <> show f)
 
-instance forall m. MonadThrow m => RunFields m '[] where
-  type RunFieldsType m '[] = ()
-  runFields _ selection = queryError ("Query for undefined selection:" <> show selection)
+instance forall ks t m.
+         ( BuildFieldResolver m (Field ks t)
+         , MonadThrow m
+         , MonadIO m
+         ) => RunFields m (Field ks t) where
+  type RunFieldsHandler m (Field ks t) = FieldHandler m (Field ks t)
+  runFields lh selection@(AST.SelectionField (AST.Field alias name _ _ _)) =
+    let NamedFieldExecutor k mValue = buildFieldResolver @m @(Field ks t) lh selection
+    in case name == k of
+      False -> queryError ("Query for undefined selection:" <> show selection)
```

Comment:
I'm unhappy about the duplication in `RunFields` instances but I think we can reduce it a bit with helper functions.